### PR TITLE
Introduce a memo monad

### PR DESCRIPTION
**Pull request by jeremiedimino**
This PR is a first step towards a memo monad. It adds a `Memo.Build` monad that is currently just fiber, update the `Memo` API to use it rather than `Fiber.t` and update the client code. There are probably too many `Memo.Build.run`/`Memo.Build.of_fiber` but it's a start.

### CODE REVIEW

**Comment by aalekseyev** [comment]
I feel like there's a significant amount of code updated in this PR that should not be in the `Memo.Build` monad.
We want code in this monad to be safe to memoize, so anything that looks imperative is suspicious.

For example these things don't make sense to me in the memo build monad, philosophically:

- `Artifact_substitution.copy`
- `Watermarks.subst`
- `Build_system.set_vcs`
- `do_build` calling the end-of-build hook

It seems a bit surprising that these are in Memo.Build monad, even though that's somewhat less clear:
- `Main.init_build_system`
- `Scheduler.go` and most of the stuff in `/bin`

These primitives in memo itself I find suspicious, too:

- `finalize`
- `fold_errors`

Finalize because it, obviously, means that there's some state management going on, which will probably be broken by memoization.
Fold_errors - I'm not sure, but we should have a separate discussion about error handling at some point.

A bunch of `Process.run` calls there, e.g. calls to opam, are surely not safe to memoize as written, but philosophically they are probably OK because they can be made safe by declaring some dependencies.

**Comment by jeremiedimino** [comment]
Agreed for `finalize` and `fold_errors`, and in fact all the exception related functions. However, removing them will require quite a few more changes, so I'm thinking of doing them in subsequent PR.

For `Artifact_substitution.copy`, it's a bit tricky; some substitutions require things such as `Vcs.describe` that we currently memoize. Maybe `Memo` is not the right fit for these, but we do need some way to cache/update them. But thinking about it again, we are not relying on memoization here in the same way as we do for the build. More precisely, if `Vcs.describe` change we are not going to redo the promotion. It is more a matter of: whenever the promotion happens, we capture whatever the value is at this point.

I need to have a look for the other ones. But I'm expecting that changing them will require deeper code changes. For instance, eventually we probably want to create the build system only once rather than re-create it for each new build.

**Comment by jeremiedimino** [comment]
For opam calls, in the end it's quite likely we want them to be "evaluate once at initialization" kind of calls. It's going to be impossible to declare the deps. Plus we can consider that changing the opam root is a heavy enough operation that it requires a restart of Dune.

**Comment by jeremiedimino** [comment]
Thinking out loud; it seems that we have three kinds of computations:

- restarting computations: when one of the dep change, we restart the computation (the main build)
- snapshot computations: whenever we happen to evaluate them, we want an up-to-date value and we also want caching. In between that, we don't care about changes to such values
- initialisation computations: we want to evaluate them once at initialisation time and remember the value forever

File: src/dune_engine/action_builder.ml [hunk 1]
```diff
@@ -27,8 +27,8 @@ type 'a t =
   | Memo : 'a memo -> 'a t
   | Catch : 'a t * 'a -> 'a t
   | Deps : Dep.Set.t -> unit t
-  | Fiber : 'a Fiber.t -> 'a t
-  | Dyn_fiber : 'a Fiber.t t -> 'a t
+  | Memo_build : 'a Memo.Build.t -> 'a t
+  | Dyn_fiber : 'a Memo.Build.t t -> 'a t
```
Comment (snowleopard) [inline src/dune_engine/action_builder.ml:31]: `Dyn_fiber` should probably get renamed to `Dyn_memo_build`.

File: src/dune_engine/action_builder.ml [hunk 2]
```diff
@@ -392,18 +392,18 @@ module Expert = struct
   let build f = Build f
 end
 
-let fiber f = Fiber f
+let fiber f = Memo_build f
```
Comment (snowleopard) [inline src/dune_engine/action_builder.ml:395]: Seems like one more renaming to do.

File: src/dune_engine/action_builder.ml (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_engine/action_builder.ml]: Indeed, done

File: src/dune_engine/action_builder.ml (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/action_builder.ml]: Done as well

**Comment by jeremiedimino** [comment]
I also renamed `Expert.build` to `Expert.action_builder`. Though, I'm wondering if it shouldn't be called `Expert.stage`

**Comment by snowleopard** [comment]
> * restarting computations: when one of the dep change, we restart the computation (the main build)
> * snapshot computations: whenever we happen to evaluate them, we want an up-to-date value and we also want caching. In between that, we don't care about changes to such values
> * initialisation computations: we want to evaluate them once at initialisation time and remember the value forever

@jeremiedimino I don't understand the difference between the last two categories.

You say "we don't care about changes to such values" and I find that strange. It seems to me that it's not about "not caring about these changes" but rather about "not being notified about these changes", which I presume is the case with the VCS stuff. Do you agree that ideally we want the set of such snapshot computations to be empty?

By the way, to which category would you put the computation that returns the current time (for example, to timestamp a binary)? I'd put it into the initialisation category but the current time is very much snapshot-like, so I'm not sure.

**Comment by aalekseyev** [comment]
> Thinking out loud; it seems that we have three kinds of computations:
> 
> * restarting computations: when one of the dep change, we restart the computation (the main build)
> * snapshot computations: whenever we happen to evaluate them, we want an up-to-date value and we also want caching. In between that, we don't care about changes to such values
> * initialisation computations: we want to evaluate them once at initialisation time and remember the value forever

In jenga model we don't need to distinguish between restarting and snapshot computations: both are simply Tenacious.t, the difference is that each snapshot computation is a separate call to `Tenacious.exec`, while the "restarting" computations are forced from the "main" `Tenacious.exec`. I like that approach. (there are not many snapshot computations in jenga, but you could call e.g. [jenga diagnostics build-goal] one)

Initialization computations are, I think, a form of snapshot, so in principle don't seem very different.
What may make them different is that we're likely to use them in other restarting computations, which may have an unfortunate effect of making those restarting computation operate on old values.

**Comment by jeremiedimino** [comment]
I would definitely put timestamp computation (for the purpose of capturing build info when linking a binary) in the snapshot category. In fact, for timestamp specifically we don't even need caching. Taking the current time every time seems fine.

> Do you agree that ideally we want the set of such snapshot computations to be empty?

Well, I agree for the main build. i.e. we don't want a build artifact to somehow depend on a snapshot computation. In our case, we only depend on these values for the purpose of rewriting artifacts during promotion or installation, i.e. moving things out of the build directory, so that seems fine.

> You say "we don't care about changes to such values" and I find that strange. It seems to me that it's not about "not caring about these changes" but rather about "not being notified about these changes", which I presume is the case with the VCS stuff. 

That's right.

> In jenga model we don't need to distinguish between restarting and snapshot computations: both are simply Tenacious.t, the difference is that the result for a snapshot computation is collected with a separate call to Tenacious.exec, while the "restarting" computations are forced from the "main" Tenacious.exec. I like that approach. (there are not many snapshot computations in jenga, but you could call e.g. jenga diagnostics build-goal one)

I like this as well.

> Initialization computations are, I think, a form of snapshot, so in principle don't seem very different.

Indeed. A few ideas to represent them:

- a "freeze" operation `'a Memo.Build.t -> 'a Memo.Build.t`. The result becomes a forever constant and we can forget whatever dependencies we recorded for it
- same as above but taking a fiber: `'a Fiber.t -> 'a Memo.Build.t`, to make it clearer the computation is outside of the scope of memo
- or something similar to the `Force_at_initialisation` monad we have in jenga

Thoughts?

> What may make them different is that we're likely to use them in other restarting computations, which may have an unfortunate effect of making those restarting computation operate on old values.

Indeed, but I guess it's difficult to avoid. We could also re-run such computations periodically as an approximation. Though in the case of opam I feel like it's overkill.

**Comment by snowleopard** [comment]
Just a general comment: by keeping the design/API as close to Jenga as possible (without sacrificing performance too much), we can probably simplify our subsequent rule migration.

**Comment by aalekseyev** [comment]
> A few ideas to represent [initialization computations]:
> 
> 1. a "freeze" operation `'a Memo.Build.t -> 'a Memo.Build.t`.
> 2. same as above but taking a fiber: `'a Fiber.t -> 'a Memo.Build.t`
> 3. or something similar to the `Force_at_initialisation` monad we have in jenga

I like (2) better than (1). Tenacious has a `lift` and `embed` to embed the computations in Async monad, which are in similar spirit.
`Force_at_initialisation` is not really intended to "forget" dependencies, it merely controls the evaluation order, so even if I can imagine a similar API, the goal is completely different.

**Comment by jeremiedimino** [comment]
Alright, so following this conversation, I'm going to:

- update `Artifact_substitution.copy` (and perhaps `watermarks` as well while I'm at it) to use fiber
- introduce a freeze `'a Fiber.t -> 'a Memo.Build.t` function and use it around a few opam calls

I'll leave some of the other changes such as `Main.init_build_system` for subsequent PR, as this will likely require more code re-organisation

**Comment by jeremiedimino** [comment]
So while modifying the `opam config var <var_name>` call, I realised that I needed to cache one value per `(environment, var_name)`. So I just used a memoized function that takes the environment and variable as input and it behave exactly the way we want.

I also realised that `freeze` is basically `Memo.Lazy.Async.create (fun () -> Memo.Build.of_fiber f)`. From the point of view of memo, such as node as no dependency, so it will effectively be a constant.

**Comment by jeremiedimino** [comment]
I moved the calls to opam in an `Opam` submodule and wrapped them in memoized functions.

**Comment by jeremiedimino** [comment]
While looking at this code, I was wondering about the following:

all the memoized functions take an `Env.t` structure as input, which can potentially be a big structure. We will need to traverse this structure at every call for hashing and comparison purposes. At the same, if we were issuing a bunch of opam calls in the same context, they would all use the same `Env.t` value. Which made me wonder if there shouldn't be a way for "factorise" one input for a bunch of memoized functions. In this particular case, it doesn't seem worth it though as we don't call opam many times.

**Comment by aalekseyev** [comment]
You can always do a transformation along these lines:

given 
```
f : env -> a -> b
g : env -> c -> d
```

you can construct:
```
f_and_gg : env -> { 
  f : a -> b; 
  g : c -> d;
}
```

Assuming all functions involved are memoized, you will save a lot on env comparisons.
The above of course is sometimes undesirable because it creates some coupling because the f and g parts of code base.

Another possible solution is to make env comparison cheap by introducing the hash or id.

**Comment by jeremiedimino** [comment]
At the moment this is not safe though, right? i.e. memoising a function in the middle of the execution.

**Comment by aalekseyev** [comment]
I think that's completely fine and we do it in a bunch of places (not that the latter is proving anything, hah).
A memoized function can be treated as a value in its own right and I believe can be safely cached.

**Comment by aalekseyev** [comment]
There's something you lose when you do this transformation, though: it's the ability to give memoized functions global names and therefore the ability to e.g. save/load them to/from disk.

**Comment by jeremiedimino** [comment]
Indeed. Although, in theory it feels like this could work as well.

BTW, from my point of view this PR is ready.

**Review by aalekseyev — APPROVED** [review]
The last patches look good too.

**Comment by aalekseyev** [comment]
> in theory it feels like this could work as well.

I guess to make it possible to give names to "nested" memoization cells/tables we'd need to track what (could have) went into the closure of such a function. That indeed doesn't sound impossible.

**Pull request merged**